x-pack/legacy/plugins/xpack_main/server/lib/xpack_info.ts

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { createHash } from 'crypto';
+import { Legacy } from 'kibana';
+
+import { XPackInfoLicense } from './xpack_info_license';
+
+import { LicensingPluginSetup, ILicense } from '../../../../../plugins/licensing/server';
+
+export interface XPackInfoOptions {
+  clusterSource?: string;
+  pollFrequencyInMillis: number;
+}
+
+type LicenseGeneratorCheck = (xpackInfo: XPackInfo) => any;
+
+export interface XPackFeature {
+  isAvailable(): boolean;
+  isEnabled(): boolean;
+  registerLicenseCheckResultsGenerator(generator: LicenseGeneratorCheck): void;
+  getLicenseCheckResults(): any;
+}
+
+interface Deps {
+  licensing: LicensingPluginSetup;
+}
+
+/**
+ * A helper that provides a convenient way to access XPack Info returned by Elasticsearch.
+ */
+export class XPackInfo {
+  /**
+   * XPack License object.
+   * @type {XPackInfoLicense}
+   * @private
+   */
+  _license: XPackInfoLicense;
+
+  /**
+   * Feature name <-> feature license check generator function mapping.
+   * @type {Map<string, Function>}
+   * @private
+   */
+  _featureLicenseCheckResultsGenerators = new Map<string, LicenseGeneratorCheck>();
+
+  /**
+   * Set of listener functions that will be called whenever the license
+   * info changes
+   * @type {Set<Function>}
+   */
+  _licenseInfoChangedListeners = new Set<() => void>();
+
+  /**
+   * Cache that may contain last xpack info API response or error, json representation
+   * of xpack info and xpack info signature.
+   * @type {{response: Object|undefined, error: Object|undefined, json: Object|undefined, signature: string|undefined}}
+   * @private
+   */
+  private _cache: {
+    license?: ILicense;
+    error?: string;
+    json?: Record<string, any>;
+    signature?: string;
+  };
+
+  /**
+   * XPack License instance.
+   * @returns {XPackInfoLicense}
+   */
+  public get license() {
+    return this._license;
+  }
+
+  private readonly licensingPlugin: LicensingPluginSetup;
+
+  /**
+   * Constructs XPack info object.
+   * @param {Hapi.Server} server HapiJS server instance.
+   */
+  constructor(server: Legacy.Server, deps: Deps) {
+    if (!deps.licensing) {
+      throw new Error('XPackInfo requires enabled Licensing plugin');
+    }
+    this.licensingPlugin = deps.licensing;
+
+    this._cache = {};
+
+    this.licensingPlugin.license$.subscribe((license: ILicense) => {
+      if (license.isActive) {
+        this._cache = {
+          license,
+          error: undefined,
```

Comment:
Is this branch really needed? I would expect that `license.error` would be undefined when `license.isActive` is true anyways.

Reply:
Yes, but this is `this.cache._error`, which is used by the legacy xpack plugin. Although it's not 1:1 mapping. The legacy plugin updated error whenever fetcher throws
https://github.com/elastic/kibana/pull/52507/files#diff-949ac00938b10918e9232965106cd7e1L186